R/roclet_rdx.R

```diff
@@ -0,0 +1,585 @@
+#' Roclet Extending the Standard rd Roclet
+#'
+#' This roclet extends the standard `rd` roclet by allowing
+#' - to add permitted values and default values to the `@param` tag and
+#' - to add a caption and a description to examples.
+#'
+#' The following tags are supported:
+#'
+#' - `@permitted`: Permitted values for the argument. Permitted value description
+#'     which are used for several arguments/functions can be stored in
+#'     `inst/roxygen/rdx_meta.R`. For example:
+#'     ```
+#'     list(
+#'       rdx_permitted_values = list(
+#'         mode = "`\"first\"`, `\"last\"`",
+#'         msg_type = "`\"none\"`, `\"message\"`, `\"warning\"`, `\"error\"`"
+#'       )
+#'     )
+#'     ```
+#'     The reference to the permitted values is done by specifying the name of
+#'     the list element in square brackets, e.g., `@permitted [mode]`.
+#' - `@default`: Default value for the argument. By default the default value
+#'   from the function formals is displayed. This can be overwritten by using
+#'   the `@default` tag.
+#'
+#' - `@examplesx`: This tag can be used to mark the beginning of the examples
+#'    section but doesn't affect the output, i.e., it can be omitted.
+#'
+#' - `@caption`: Caption for the example. The caption is displayed as a subsection
+#'   in the examples section. The caption can be followed by an arbitrary number
+#'   of `@info` and `@code` tags.
+#' - `@info`: Description of the example.
+#' - `@code`: Code of the example.
+#'
+#'      By default, any warning or error issued by the example code causes the
+#'      building of the documentation to fail. If this is expected, the condition
+#'      can be added to the `expected_cnds` option of the `@code` tag. E.g.,
+#'      ```
+#'      @code [expected_cnds = "warning"]
+#'      ```
+#'
+#' To use the roclet call `roxygen2::roxygenise(roclets =
+#' "admiral::rdx_roclet")` or add to the `DESCRIPTION` file:
+#' ```
+#' Roxygen: list(markdown = TRUE, roclets = c("collate", "namespace", "admiraldev::rdx_roclet"))
+#' ```
+#'
+#' @keywords documentation
+#'
+#' @export
```

Comment:
It would be cool to showcase the extra tags in an example section here, but not sure if that even makes sense or is possible?

Reply:
Yes, that would be cool. But I'm not sure how to do it in a well readable way. We could add an example like:
```
> roxygen2::roc_proc_text(
+     rdx_roclet(),
+     c(
+         "#' Testing",
+         "#' @param x An argument",
+         "#' @permitted A number",
+         "test_fun <- function(x) 42"
+     ))
$test_fun.Rd
% Generated by roxygen2: do not edit by hand
% Please edit documentation in ./<text>
\name{test_fun}
\alias{test_fun}
\title{Testing}
\usage{
test_fun(x)
}
\arguments{
\item{x}{An argument

\describe{
\item{Permitted values}{A number}
\item{Default value}{none}
}}
}
\description{
Testing
}
```
What do you think?

Reply:
Yes, I was envisaging something like that - could you pls put in the PR so we can render and review?

Reply:
I've extended the example a bit such that all custom tags occur.

Reply:
What about just making a vignette for this and linking to that in our function documentation and in the programming strategy.  feels like we need more room!